### PR TITLE
[CI] Check test if the `GenerationTesterMixin` inheritance is correct 🐛 🔫 

### PR DESCRIPTION
**Pull request by gante**
# What does this PR do?

If a model inherits from `GenerationMixin`, then it can call `generate`. In those models, `GenerationTesterMixin` must be added to the tester in order to test `generate`. Sometimes we forget to do it. Not anymore 😈 

This PR follows up on #33212 and removes the last piece of common human error. It adds a meta-test to ensure we have the correct inheritance, which enables us to confirm that the right tests are being run automatically. In other words, if the model inherits from `GenerationMixin`, this test confirms that the tester inherits from `GenerationTesterMixin` (and the other way around too). This test also disincentivizes complex tester inheritance hierarchies, as in `idefics`, `bark`, or `musicgen`, which hurt long-term maintenance. 

In the process, it uncovered \*many\* cases of incorrect/missing inheritance, which are dealt with in this PR. Look at all these models that should have been tested with `generate` and haven't been since they were added 😭 

👉 reviewer: please start by reviewing `tests/test_modeling_common.py` and `modeling_utils.py`, then check the other files.

### CODE REVIEW

File: tests/generation/test_utils.py [hunk 1]
```diff
@@ -2753,6 +2753,41 @@ def test_speculative_sampling_target_distribution(self):
         self.assertTrue(last_token_counts[8] > last_token_counts[3])
 
 
+global_rng = random.Random()
```
Comment (gante) [inline tests/generation/test_utils.py:2756]: copied from `test_modeling_common`, otherwise we would have circular dependencies

Comment (zucchini-nlp) [reply tests/generation/test_utils.py:2756]: a comment with "copied from" can be added i think

File: src/transformers/modeling_utils.py [hunk 2]
```diff
@@ -1645,17 +1645,14 @@ def base_model(self) -> nn.Module:
...
 
         Returns:
             `bool`: Whether this model can generate sequences with `.generate()`.
         """
         # Directly inherits `GenerationMixin` -> can generate
         if "GenerationMixin" in str(cls.__bases__):
             return True
-        # Model class overwrites `generate` (e.g. time series models) -> can generate
-        if str(cls.__name__) in str(cls.generate):
-            return True
```
[the remark this answers is not in the document]
Comment (zucchini-nlp) [reply src/transformers/modeling_utils.py:1658]: is it completely different or uses part of `generate()`, like some audio models?

File: tests/models/big_bird/test_modeling_big_bird.py [hunk 3]
```diff
@@ -451,6 +451,8 @@ class BigBirdModelTest(ModelTesterMixin, PipelineTesterMixin, unittest.TestCase)
         if is_torch_available()
         else ()
     )
+    # Doesn't run generation tests. There are interface mismatches when using `generate` -- TODO @gante
+    all_generative_model_classes = ()
```
Comment (zucchini-nlp) [inline tests/models/big_bird/test_modeling_big_bird.py:455]: for my understanding: do we need to have empty `all_generative_model_classes` if no `GenerationTesterMixin` is added?

File: tests/models/udop/test_modeling_udop.py [hunk 4]
```diff
@@ -419,6 +420,14 @@ def test_model_from_pretrained(self):
         model = UdopForConditionalGeneration.from_pretrained(model_name)
         self.assertIsNotNone(model)
 
+    @unittest.skip(reason="TODO: Fix me @joao")
+    def test_generate_with_head_masking(self):
+        pass
```
Comment (zucchini-nlp) [inline tests/models/udop/test_modeling_udop.py:425]: head-mask test can be deleted after https://github.com/huggingface/transformers/pull/35786/ is merged

**Review by zucchini-nlp — COMMENTED** [review]
Great, thanks for aligning generation tests into one way. I left a few questions, for parts I didn't understand

File: tests/generation/test_utils.py (hunk 1 above)
Comment (ydshieh) [reply tests/generation/test_utils.py:2756]: are these used in a lot of places in this file, or just inside one method?

If so, we can probably avoid circular dependencies by importing them within that (single) method ..?

**Review by ydshieh — APPROVED** [review]
Nice!

File: src/transformers/modeling_utils.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (gante) [reply src/transformers/modeling_utils.py:1658]: it's completely different ☠️

File: tests/generation/test_utils.py (hunk 1 above)
Comment (gante) [reply tests/generation/test_utils.py:2756]: That's a good idea, moving to an internal import to prevent code bloat

File: tests/models/big_bird/test_modeling_big_bird.py (hunk 3 above)
Comment (gante) [reply tests/models/big_bird/test_modeling_big_bird.py:455]: If a model inherits `GenerationMixin`, one of these two must happen:
1. `GenerationTesterMixin` must be in the tester 
2. [when the tests are broken, this should NEVER happen on new models] we specify `all_generative_model_classes = ()` to make it very clear we're NOT running generation tests

option 2 is intentionally annoying (we are forced to overwrite a property), so we are very explicit about skipping tests. We don't want skips to happen unless we're very intentional about it.

**Comment by gante** [comment]
@zucchini-nlp PR comments addressed/replied 🤗

File: tests/generation/test_utils.py (hunk 1 above)
Comment (ydshieh) [reply tests/generation/test_utils.py:2756]: one another possible approach is not to use

> issubclass(self.__class__, GenerationTesterMixin),

but check the `__name__` of all parent classes (but recursively, but maybe no ) and see if `GenerationTesterMixin` is in that set.

Up to you: `# copied` is less maintained and we would like to get away from it (since now we rely more on modular)

Comment (gante) [reply tests/generation/test_utils.py:2756]: It's also okay to have a pure copy of the short functions :P It's just a handful of lines, I don't think it's worth the extra work for now -- I will have to refactor these lines when we remove TF (i.e. very soon) 👀

Comment (ydshieh) [reply tests/generation/test_utils.py:2756]: OK 👍

**Review by shethaadit — APPROVED** [review]

**Review by zucchini-nlp — APPROVED** [review]
Thanks!
